Include/object.h

```diff
@@ -78,12 +78,78 @@ whose size is determined when the object is allocated.
 #define PyObject_HEAD                   PyObject ob_base;
 
+/*
+Immortalization:
+
+The following indicates the immortalization strategy depending on the amount
+of available bits in the reference count field. All strategies are backwards
+compatible but the specific reference count value or immortalization check
+might change depending on the specializations for the underlying system.
+
+Proper deallocation of immortal instances requires distinguishing between
+statically allocated immortal instances vs those promoted by the runtime to be
+immortal. The latter which should be the only instances that require proper
+cleanup during runtime finalization.
+*/
+
+#if SIZEOF_VOID_P > 4
+/*
+In 64+ bit systems, an object will be marked as immortal by setting all of the
+lower 32 bits of the reference count field.
+
+i.e in a 64 bit system the reference count will be set to:
+    00000000 00000000 00000000 00000000
+    11111111 11111111 11111111 11111111
+
+Using the lower 32 bits makes the value backwards compatible by allowing
+C-Extensions without the updated checks in Py_INCREF and Py_DECREF to safely
+increase and decrease the objects reference count. The object would lose its
+immortality, but the execution would still be correct.
+
+Reference count increases will use saturated arithmetic, taking advantage of
+having all the lower 32 bits set, which will avoid the reference count to go
+beyond the refcount limit. Immortality checks for reference count decreases will
+be done by checking the bit sign flag in the lower 32 bits.
+*/
+#define _Py_IMMORTAL_REFCNT UINT_MAX
+
+#else
+/*
+In 32 bit systems, an object will be marked as immortal by setting all of the
+lower 30 bits of the reference count field.
+
+i.e The reference count will be set to:
+    00111111 11111111 11111111 11111111
+
+Using the lower 30 bits makes the value backwards compatible by allowing
+C-Extensions without the updated checks in Py_INCREF and Py_DECREF to safely
+increase and decrease the objects reference count. The object would lose its
+immortality, but the execution would still be correct.
+
+Reference count increases and decreases will first go through an immortality
+check by comparing the reference count field to the immortality reference count.
+*/
+#define _Py_IMMORTAL_REFCNT (UINT_MAX >> 2)
+#endif
+
 #define PyObject_HEAD_INIT(type)        \
     { _PyObject_EXTRA_INIT              \
     1, (type) },
 
+#define _PyObject_HEAD_IMMORTAL_INIT(type)      \
+    { _PyObject_EXTRA_INIT _Py_IMMORTAL_REFCNT, type },
+
+// TODO(eduardo-elizondo): The change to `PyVarObject_HEAD_INIT` is
+// temporary to ease up code review. There are currently ~600
+// references to `PyVarObject_HEAD_INIT` that have to be replaced
+// with PyVarObject_HEAD_IMMORTAL_INIT as all the use cases here
+// are in relation to static objects, which are safe to mark as
+// immortal. This will be done after the inital rounds of reviews.
```

Comment:
Yes, it can definitely be done in a separate PR. The main reason I did it here was to get the final perf number. I just created a new PR with this change: https://github.com/python/cpython/pull/103308

Reply:
I closed the other PR and instead integrated the changes here with the `Py_BUILD_CORE` flag. I'll try to punt the discussion of whether or not PyObject_HEAD_INIT should by default use the immortal value to some later time (outside of this PR).